### ISSUE TITLE
3.0 Feature Ideas Collection (ideas are welcome)

### ISSUE THREAD

**Issue by BioTechproject27** [agent-LuluDodo/RebindMyKeys#4]
Yea, that. It's per default F5.
Would be cool if I could just leftclick near servers to reload them, instead of having to go to the F5 key :D

**Comment by agent-LuluDodo** [comment]
Added KeyBind: https://modrinth.com/mod/rebindmykeys/version/2.3

Note that the mod doesn't support mouse buttons :( 
I'm trying to get them working, but that might take a bit. I will leave this issue open and close it once support for the mouse is available.

**Comment by agent-LuluDodo** [comment]
Created a new issue (#5) for the mouse support, so I'm going to close this one
Thanks for suggesting :)

**Comment by BioTechproject27** [comment]
Hey so I've got a few more ideas, writing them in here so u can make your own checklist thingy in "issues" if u want:
- Adding keybind for the pie chart (default `shift + f3`; renamed "profiler chart" in 1.20.2, default `1 + f3`) and debug FPS/TPS/MSPT graph (default `alt + f3`; 1.20.2 default `2 + f3`) and as of 1.20.2 the new ping and received network traffic charts (default `3 + f3`), although since they're tied to the debug (f3) button idk how realistic that is
- Seperating some keybinds (e.g. seperating narrator from `ctrl + key`, or debug from being `f3 + key` to custom key(s)-combinations) and adding the ability to make custom key combos for other buttons (wherever it makes sense)
- on that same topic, adding the ability to have "advanced" settings (ofc only whereever some of them make sense), like what is used in e.g. tweakeroo, tweakermore, itemscroller, minihud, litematica, etc:
![2024-01-22_22 07 59 Unable to determine world name   0 0 0](https://github.com/agent-LuluDodo/RebindMyKeys/assets/100373243/e4b8ddad-3f02-432f-a4fd-24232f06c207)

I think those could be pretty dope additions, but those are bigger projects ofc

**Comment by agent-LuluDodo** [comment]
**Thank you so much!!!**

I will definitely try adding these features. The current version already supports The f3 graphs in 1.20.2+, but it's a bit buggy. And you can rebind f3 (this will also change the f3 + combinations) it's called *Debug Menu* and in the misc category.

## TODO

- [ ] Debug charts in <1.20.2
- [ ] Allow custom keybind combinations
- [ ] Make vanilla and other mod keybinds use custom keybinds
- [ ] Multiple Keys to one KeyBinding
- [ ] Advanced settings

**Comment by BioTechproject27** [comment]
> And you can rebind f3 (this will also change the f3 + combinations)

Yeah, in the 2nd bullet point I meant those combinations specifically. that they are seperated from the debug menu key. E.g. having the ability to rebind "Show Advanced Tooltips" from being `debug key + other key` (default `F3 + H`) to a single key, let's say the number key 9.
And, if the "Advanced Settings" idea is also added, then one could even have multiple bindings for one key/key combo. E.g. so that by pressing the `R` key, Chunk Borders are activated as well as Showing Hitboxes. Or that they are always active when the F3 menu is also open, etc.

**Comment by agent-LuluDodo** [comment]
### If you don't mind, could you please give me some feedback on my layout for the advanced config?

What do you think of this setup? 
Is there too much going on, or is the layout confusing?

## Simple Menu
![SimpleKeyBindsScreen](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/bfa94e8e-f6d9-4bac-90a4-f980d400c46a)

## Advanced Menu
Clicking the `Advanced` button leads to this screen
![AdvancedKeyBindsScreen1](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/cab7771c-2694-47b4-8c73-0f4da2f5677b)
Clicking on `Key Combo` leads to this screen
![AdvancedKeyBindsScreen2](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/3610d3ee-b00c-4509-90eb-f2b00e36408d)
Click on `Mode` in the Screen from before leads to this screen
![AdvancedKeyBindsScreen3](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/1025b761-2609-4e67-97fe-b8c315484b04)
(This looks a bit messy, but the implementation will show the unsupported part if I didn't define a default mode for the KeyBind, to hopefully allow support with more mods)

**Again thanks @BioTechproject27 :)**

**Comment by BioTechproject27** [comment]
Hey sorry for the late reply.. Anyway, I'd love to give some input :D

Hmm, maybe it's just habit, but I liked the relatively intuitive implementation of the example mod(s).
Because as it is now it kinda has too many seperate screens for what it needs to do.

So if I were to implement this, I would maybe have a little `Cog Wheel` button infront of every `Reset` button [exactly the same as the "U" button for unbinding].
This `Cog Wheel` would display a pop-up, with similar options to the example mods.
That means:
"Activate On" `Press`/`Release`/`Both` (Should the keybind activate upon pressing the button, releasing or both)
"Context" `In-Game`/`GUI`/`Menu` (Can the keybind be activated in GUI's, only when inside a menu [e.g. Game Menu] or only when In Game)
"Always Action" `Toggle`/`Activate`/`Deactivate` (basically always only deactivates/activates/toggles keybind [this is entirely my idea, the reason for its implementation is explained in the 3rd section])
"Allow Extra Keys" `True`/`False` (wether keybind should activate when additional keys are pressed)
"Order Sensitive" `True`/`False` (wether to honor the order of keys pressed or not)
[Maybe move the `Unbind` button in here for cleaner Keybind UI?]

(Descriptions should be displayed as hover tooltips to better help ppl understand what is meant and how it works.)


I'd also add a little `+` button next to the `Cog Wheel` button for adding multiple keybinds [It would just add a new line which defaults to `Unbound` and would have its own `Cog Wheel`. Ofc it should also have a `-` or `x` for removing it].
(The idea is that one can have e.g. the Debug Menu pop up as well as showing chunk borders when pressing `9`, *and* can activate the narrator and show chunk borders when pressing `8`, while also being able to deactivate the narrator seperately by pressing `7` *and* be able to deactivate all at once by pressing `6`. This is just a wild example, but especially with other mods this becomes very helpful.)

Making and having keybinds that use multiple keys would be exactly the same as the example mods implementation (click on the button field and press the buttons you want).



Ofc the default options should be Vanilla.
e.g. for the Debug Menu that means keybind `F3`
"Activate On" `Press`
"Context" `In-Game`
"Always Action" `Toggle`
"Allow Extra Keys" `True`
"Order Sensitive" `True`/`False` (doesn't matter here as it's only one key. Could also maybe just be greyed out.)



As a design for the `Cog Wheel` pop-up, probably a vanilla container (e.g. barrel) would do with similar dimensions relative to the screen, maybe having an alpha of 90 or so. And also like with many pop-ups, clicking around it would close said pop-up and save the changes.
I think the `Reset` should also reset everything in the `Cog Wheel` to default.


These are just my ideas and I'd love to hear and discuss other ideas!
(also if anything is unclear in what I described pls ask)

**Comment by agent-LuluDodo** [comment]
Thank you so much for the feedback! I'll make a new design and send it in here once I have some time. You don't have to be sorry. It's you helping me after all :)

**Comment by agent-LuluDodo** [comment]
![new_design_preview](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/a06101a7-9483-48db-88db-32195a1c5d5b)
Here's a preview of the new design if you're interested :)
## Buttons
- Edit: Allows you to change the bound button (vanilla) and shows unbind button while capturing input (custom)
- Add: Allows you to add another KeyBinding with a remove button (custom)
- Settings: Advanced configuration for the current KeyBinding (custom) [potentially also for extra KeyBindings]
- Reset: Resets everything for the current KeyBindings, including advanced configuration and extra KeyBindings (custom + vanilla)
- Remove: Removes this extra KeyBinding (custom)
- Unbind: Unbinds the current KeyBinding (custom) [also potentially also for extra KeyBindings]

**Comment by BioTechproject27** [comment]
> Thank you so much for the feedback! I'll make a new design and send it in here once I have some time. You don't have to be sorry. It's you helping me after all :)

Well, it's not like I'm not getting anything out of this myself. Still ghosting people kinda sucks bcs then they'll always wonder what went wrong xd

> Settings: Advanced configuration for the current KeyBinding (custom) [potentially also for extra KeyBindings]
Reset: Resets everything for the current KeyBindings, including advanced configuration and extra KeyBindings (custom + vanilla)

So the way it looks right now, the settings button would influence all keybindings for that action. It would be nice to have one for each keybind seperately.
The same should then be true for the reset.

I think an improvement would be to have the same buttons for each line in some order, like:
`X` (unbinds if it's the first line; removes whole keybind line if it's secondary or more) `+` (adds another keybind line) `Cogwheel` (Settings for this keybind line) `Reset` (resets keybind & keybind settings in this line to vanilla settings)

This way would let you configure each line and as a bonus it looks clean.

**Comment by agent-LuluDodo** [comment]
# Key Binds Screen
![2 0-SimpleKeyBindsScreen](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/e6159266-57c8-49f3-865b-eb8cd2b30889)
I changed the design a bit to make it less crammed. I wasn't sure if I should keep the spacing between the gear and reset button.
What should the rebind button next to the extra KeyBindings do? Should it only reset the settings or also set the key to *unbound* (removing the row would be the same as the remove button)?

## Unbind
The `Unbind` button could be in a right click menu, but I don't think that looks nice after trying it:
![2 0-SimpleKeyBindsScreen Right Click](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/06e966aa-7e12-4ec6-878e-fd9e7223eb05)
I could imagine using the previous method of it showing when rebinding the key, or I could move the `+` button, but I don't know where I should put it.

## Settings
After clicking on the `gear` this popup opens:
![2 0-SimpleKeyBindsScreen Settings](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/e1f0c75f-c98a-4f49-9192-23460e0fa21a)
The Context is read-only, since it doesn't make sense to be able to change that. I also renamed *Always Action* to *Operation Mode*, because I think it fits better.

### Key Combination
After clicking on `Edit` in the Key Combination row, this menu opens:
![2 0-SimpleKeyBindsScreen Key Combination empty](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/1d4ead8d-b2cd-4154-a7a8-4a031ac00c75)
Should I replace the `Edit` button in the settings screen with this list? That would reduce the amount of screens to 2.
Pressing the `Add` button opens a dropdown for more options (see next picture). The KeyBinding visible in this picture is the one from the original menu. 

A KeyBinding is structured like this:
 1. Edit Button
 2. Remove/Unbind Button
 3. Key this results in
 
`Order Sensitive` can be one of two values:
 - Order Sensitive | Requires all extra KeyBindings to be pressed at the time of pressing the main KeyBinding. (default)
 - Ignore Order |  Allows you to press the extra KeyBindings after pressing the main KeyBinding.
 
Here is a picture with more going on to explain the last feature:
![2 0-SimpleKeyBindsScreen Key Combination](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/542f1a13-a1a6-43d9-a961-759e3a459059)

The `Add` button dropdown shows 3 options:
 1. Key | Adds a normal KeyBindings which is bound to the next key/mouse button you press.
 2. Key Binding | Requires the KeyBinding specified to be active.
 3. Modifier | These can be SHIFT, CTRL, ALT, SUPER ([GLFW docs](https://www.glfw.org/docs/3.3.2/group__mods.html))
 
If any feature is unclear, or you have more improvements to suggest, I would appreciate that :)
Also, please let me know if I should combine the *settings* and *key binding* menus.
 
**As always thanks for all your help!**

**Comment by BioTechproject27** [comment]
> I wasn't sure if I should keep the spacing between the gear and reset button.

Hmm, maybe add spacing between `Key` and all the other buttons instead? I think because `Key` is so wide you automatically get this crammed look.

> What should the rebind button next to the extra KeyBindings do? Should it only reset the settings or also set the key to unbound (removing the row would be the same as the remove button)?

I've got another idea here, I think the `Reset` button should stay as is and reset everything for that Action to default (preferably with a confirmation screen).
Resetting to the default option for each keybind line could be done in the `Gear` pop-up, maybe with your compact reset button design in the upper right/left corner?
It should also only reset the options in the `Gear` then, as one can simply remove and add a new line if they'd want a full reset by pressing `X` and `+`.

On the topic of unbinding keys, the way I see it, unbinding a button only really makes sense when you simply don't want that action ever performed by a key. So unbinding should only be an option for the first (and only) line.
I could also be wrong ofc, but I don't see a scenario where one would need an unbound key **and** have additional keybind lines.

> The Context is read-only, since it doesn't make sense to be able to change that.

For Vanilla settings I absolutely agree, mod support can be added later and is a whole different can of chaos for now xD

> I also renamed Always Action to Operation Mode, because I think it fits better.

Yep, agree.




Okay, so. In terms of actually binding a key(combo) I'd refer to the example mods (e.g. [tweakeroo](https://www.curseforge.com/minecraft/mc-mods/tweakeroo)). Have you played around with the keybinding implementation they have? Highly reccommend checking it out. It's very clean and intuitive and removes the need for its own screen as far as I see.
Instead of having Modifiers & Keys, it just treats all of them as Keys.

But I like your idea of checking for active keybinds, I'd move that (and order sensitivity in that case as well) into the `Gear` if we remove that keybindings screen entirely. Maybe at the bottom of the pop-up (which would probably need to be scrollable) as "Active Keybinds":
```
Action_1 - Keybind(combination)_1
```
```
Action_2 - Keybind(combination)_2
```
etc.
Maybe with a `+` and `X` button behind each to add/remove new lines.
As I am writing this I again realize I'm just blatantly plagiarizing a feature of tweakeroo & similar mods, specifically the "List" feature lmfao. But hey, it works pretty well.
The newly added line (with `+`) should be empty though in that case and have a dropdown of all available keybindings.

**Comment by BioTechproject27** [comment]
> The Context is read-only, since it doesn't make sense to be able to change that.
For Vanilla settings I absolutely agree, mod support can be added later and is a whole different can of chaos for now xD

After looking at the keybinds there are a few exceptions I thought of. Those would be:


**Movement**
Sneak (could be usable in GUIs & should be able to restrict it in game)
Sprint (could be usable in GUIs, makes sense when it's set to toggle & should be able to restrict it in game)

**Gameplay**
Dismount (could be usable in GUIs & should be able to restrict it in game)

**Multiplayer**
List Players (could be usable in GUIs & should be able to restrict it in game)

**Miscellaneous**
Debug Menu & everything attached to it (could be usable in GUIs & should be able to restrict it in game)
Narrator (should be able to restrict it)
Toggle HUD (could be usable in GUIs & should be able to restrict it in game)
Toggle Perspective (could be usable in GUIs  & should be able to restrict it in game)

[Technically the other **Movement** & **Gameplay** keys could also be usable in GUIs, but I feel like that's cheaty and hack clients already do that.]

**Comment by agent-LuluDodo** [comment]
Just wanted to give an update. I've got the foundation down, but still need to implement a lot. Since exam period is coming up for me, that might sadly take a while. Thanks for your patience :)

**Comment by BioTechproject27** [comment]
Ah, well I was actually gonna ask for an update on how the project's goin' :>
No worries, school & jobs come before hobbies!
An update would be nice when you continue working on the project.
And good luck on your exams! :>

**Comment by BioTechproject27** [comment]
Additional idea:
Fast succession button pressing & customization
- Essentially, adding the option to add a double/triple/etc tap option with configurable min & max tick delay
(the latter would essentially customize the wiggle room for failure of tapping fast, especially nice for people with impaired motor functions).

Reasoning: Double tapping `jump` allows for creative flight. Adding this tapping option would also allow to freely configure this (and could allow for other stuff, like disabling/rebinding double tapping the `forward` key which activates sprinting as a secondary option and is permanently tied).

**Comment by BioTechproject27** [comment]
I browsed through the [minecraft wiki](https://minecraft.wiki/w/Controls), where I found that "`F4` Toggles the currently active post-processing shader, such as screen filters when spectating as certain mobs."

This is seperate from the gamemode changer and should also be added for completion.
(This is a feature, while only used in vanilla for spectating mobs afaik, is used by other mods like the [Origins](https://modrinth.com/mod/origins) mod)

**Comment by agent-LuluDodo** [comment]
That's what F4 does, I saw it in code but pressing f4 didn't do anything, so I just ignored it. Adding it to my ToDo

**Comment by agent-LuluDodo** [comment]
Managed to make a bit of progress here and there. So here is a little sneak peek of my current progress :)
![Screenshot 2024-05-10 221925](https://github.com/agent-LuluDodo/RebindMyKeys/assets/98981925/13422dee-9a7d-4ee7-987f-3c63266d5e53)

**The main points I still need to do are:**
- finishing the gui (add button, popups)
- saving
- context changing
- order sensitivity
- double, triple tapping
- some bug-fixes

**Comment by BioTechproject27** [comment]
Hey, how is it comin' along? c:

Had some other ideas, one which goes kinda in the same direction:

Adding control to other aspects that lack them:

Toggle Auto Jump
Toggle Subtitles
Decrease/Increase/set specific value Brightness
Decrease/Increase/set specific value FOV
Decrease/Increase/set specific value Sound Volume
Toggle View Bobbing
seperating the Debug bound keybinds (like showing chunks, hitboxes, etc) (just mentioned it in case it wasn't on the list yet)
and maybe **Skin Settings**:
Toggle Cape
Toggle Hat
Toggle Jacket
Toggle Left Pants Leg
Toggle Left Sleeve
Toggle Right Pants Leg
Toggle Right Sleeve

Another idea would be, if u don't have it already, but a confirmation screen for resetting the keybinds makes sense.

And, another feature from other mods I found which could make sense:
profiles/instances/presets for keybinds

And modding for 1.21 makes sense as it's out now, plus, it may be easier with Quilt instead of fabric? At least they should be pretty compatible afaik

Anyway, let me know ur thoughts sometime

**Comment by agent-LuluDodo** [comment]
Really thankful for all the suggestions but I'm currently on holidays so give me till Sunday for the reply. :)

**Comment by BioTechproject27** [comment]
just answer whenever, this is not time sensitive lmao

**Comment by agent-LuluDodo** [comment]
Don't worry, I'm not stressing to answer these comments. I'm just really invested into this project. kekw

> Hey, how is it comin' along? c:

Taking longer than expected, but I also always underestimate the amount of work for projects. xD

I'm currently working on saving and profiles. I've finished most of the essential features (contexts, key-combos, multi clicks). Still need to test order-sensitivity and fix a bunch of bugs.

Here's a picture of the current menu:
![2024-07-21_16 04 54](https://github.com/user-attachments/assets/a5409bab-cca2-4253-9680-84bd59e59705)
You might have noticed that the + button is missing. I can't think of a good spot for it, do you have an idea?

Also planing on releasing a beta version soonish.

> Adding control to other aspects that lack them:

That's an amazing idea, definitely adding those.

> Another idea would be, if u don't have it already, but a confirmation screen for resetting the keybinds makes sense.

Also adding the confirmation screen (you're going to be able to disable it in the settings).

> seperating the Debug bound keybinds (like showing chunks, hitboxes, etc) (just mentioned it in case it wasn't on the list yet)

Already on the list ;)

> profiles/instances/presets for keybinds

Already working on those. Do you have an idea for how to switch between profiles?

> And modding for 1.21 makes sense as it's out now

It's easier to finish the mod in 1.20.4 and then port to 1.21, since I'll be done soon™.

> plus, it may be easier with Quilt instead of fabric? At least they should be pretty compatible afaik

Fabric mods should be able to run on Quilt if you install [Quilted Fabric API](https://modrinth.com/mod/qsl), but I haven't tried it. As for switching to Quilt, even if they are similar, there are still some differences, so I would have to redo parts of the mod. Might check out Quilt if I do a new project, but switching RebindMyKeys to Quilt isn't going to happen (at least for 3.0).

**Comment by BioTechproject27** [comment]
> but I also always underestimate the amount of work for projects

real xd

> \+ button is missing. I can't think of a good spot for it

perhaps move the `X` [which I assume simply unbinds] into the key-settings-menu/cogwheel and have the `+` at that position instead?
But tbh, I think _where_ they are isn't as important as getting the features working xd I think the buttons can be shuffled around later after an initial 'alpha'/'beta' build or sth

Btw, just outta curiosity, what are those symbols on the far left? c:

> but switching RebindMyKeys to Quilt isn't going to happen (at least for 3.0)

Fair.
My thinking was that (which I should've mentioned xd), depending on how the mod works, Quilted Fabric API may not be enough, bcs as I see it it's just a 'translator', and those often suck up performance, which _could maybe_ have an impact, esp on weaker machines, bcs this mod does something every time a button is pressed.

So like, when considering other platforms, _optimization_ is often neglected (which is why so many mods and games in general, often suck up resources like crazy). Just sth that should be kept in mind :3

**Comment by agent-LuluDodo** [comment]
> the `X` [which I assume simply unbinds]

To be exact, the `X` removes the current row, but that's basically the same. It rather keep it accessible, since adding and removing combos are the most important features imo.

>  I think the buttons can be shuffled around later after an initial 'alpha'/'beta' build or sth

Yeah, imma just put it somewhere until I find a better place for it :)

> Btw, just outta curiosity, what are those symbols on the far left? c:

They show the "mode" the keybinding is in (you can change the mode in the settings).
- ![Action Mode Icon](https://github.com/user-attachments/assets/65e53ac8-304f-4794-91de-69c9a0fb5f6d) Is the `Action` Mode that means pressing/releasing the button triggers it.
- ![Hold Mode Icon](https://github.com/user-attachments/assets/afb27d7d-7c47-424e-9e8b-086185b5c5c6) Is the `Hold` Mode that means the button is active while being held down.
- ![Toggle Mode Icon](https://github.com/user-attachments/assets/cdb96b64-092f-433b-95fd-55142c8562db) Is the `Toggle` Mode that means the button switches between being active and inactive on press.

*I'm also going to add toolbinds to them which will have the name of the mode, since remembering the icons could be annoying.*

> Quilted Fabric API may not be enough, bcs as I see it it's just a 'translator', and those often suck up performance, which _could maybe_ have an impact, esp on weaker machines, bcs this mod does something every time a button is pressed.

**TLTR:** Quilted Fabric API shouldn't cause performance issues.

This might be a bit technical, cause I suck at explaining stuff, so here it goes:
I'm only using Fabric API in a few places (most of it for translations, which iirc work the same way in Quilt). Something like button presses have nothing to do with the API since they are handled directly by minecraft or my code. Fabric API was designed for compatibility between mods (such as translations, commands, drawing huds, etc..) (Called an abstraction layer). 

The bigger problem are my mixins, which might just not work with Quilt.

> _optimization_ is often neglected

That's true, but my aim is to provide a better keybind system, so I'm trying to improve latency (the time between pressing a key and the game registering the press). But I'm not experienced in this, so it might end up slightly worse than vanilla. The mod shouldn't drop your fps at all and if that ends up happening, please create an issue, so I can fix it. (This is for both Quilt and Fabric)

I actually have some ideas for performance optimizations which I'm going to try and implement later on.

*I'll make a port for Quilt if it turns out that the mod doesn't work or causes performance issues with Quilt.*

---

**PS:** If want more performance and a have compatible GPU (Nvidia 16xx, 20xx, 30xx or 40xx series) try [Nvidium](https://modrinth.com/mod/nvidium). I assume u already have [Sodium](https://modrinth.com/mod/sodium), [Lithium](https://modrinth.com/mod/lithium) and [Entity Culling](https://modrinth.com/mod/entityculling). :]

**Comment by agent-LuluDodo** [comment]
I have released an alpha you can try! :)

### There are still a lot of bugs I need to fix, so don't expect everything to work perfectly!
If you encounter any bugs, please let me know, though.
Also note that configuration files from alpha won't work with the finished version.
Menus also currently look terrible, since I just threw every option I could think of in them.
Forth note, other mods won't work with this since I haven't implemented mod support.

[RebindMyKeys 3.0 Alpha 3](https://github.com/agent-LuluDodo/RebindMyKeys/releases/tag/3.0%2Balpha3)

Here is a list of currently known bugs:
- Global settings always asking to save
- Reset doesn't reset keys
- Left click not registering after recording key
- Debug charts don't work
- Debug Menu with release doesn't work
- Profiles sometimes failing to delete
- Sliders don't work

Also small explanation of the main menu:
![main menu](https://github.com/user-attachments/assets/e07a7f36-7126-44c3-aa1c-6719b041162a)

**Comment by BioTechproject27** [comment]
Oh hey, sorry it took so long to reply...
(I thought I did to the msg in august, but looks like i forgot >->)

>my aim is to provide a better keybind system, so I'm trying to improve latency

Ooooh niceee, definitely worth a shot
Also, I was not necessarily talking about my machine (it's more like a low, maybe a mid-tier one, but works fine for minecraft), just more generally xd
And yeah, I also have mods like ferrite core, lazy dfu, etc etc

Aaaaaanywayyyyy

It's already pretty cool :D
Buttons already pretty intuitive c:
(e.g. I didn't look at your picture, saw the stacked ^ and thought "Is that collapse all? Niceee")
Immediately bound jump to left-click, which was pretty funny ngl

And myep, found things I'd call bugs:
- Weird behavior with the mouse buttons, e.g. binding left-click to jump works fine, no complaints there, but then when adding other buttons like `space` makes them both not work. To get it to work, I have to add another button (like `N`), press `Done` to exit the menu, and then delete that extra button (`N`) & press `Done` again and then it'll work lmao
And after _that_, you can change the leftclick to anything and back and it'll work properly, unless you delete the left-click keybind, then you gotta redo the steps xd
(I love finding bugs, it's quite honestly fun witnessing such weird behaviour and figuring out specifics lmao)
Anyway, that's likely related to what you meant:
`Left click not registering after recording key`
So ye, it does register it, but only after that ritual.

- `Reset doesn't reset keys` Yupp, can confirm
- `Debug charts don't work` Yep, F3 just closes the debug menu, and setting as seperate keys don't seem to do anything
- `Sliders don't work` Slider not sliding when I clickie, yep
- Am guessing "Key Combination" is still very much WIP since just a menu opens and nothing i click seems to change anything
- Idk why, but my mmb functionality inside inventories is gone, yoinked, even (mainly creative stuff like stack grabbing and dragging to fill inventory slots)
- Gamemode switcher stops showing the pop-up if other keys are used

Didn't mention tiny bugs, like
e.g. After 'Recording' a button press, the next virtual button (like `Done` the cogwheel or `X`, or even when entering the cogwheel menu using tab & enter and then wanting to close the menu with `Quit`) needs to be left-clicked twice for it to do.. idk, its function when clicked yk?
Or the buttons not really updating, e.g. after changing the number of key presses
Or stuff like asking for confirmations in the profile thingy twice n such, bcs i think stuff like that is less about ur mods core functionality and more so touch-ups for later (probably, unless one decides to become a large headache for no reason, but that's a hypothetical to worry for a later time i'd say xd)
_However_:
Funny, more so cosmetic things i just wanted to share:
Buttons highlight below the dirt-layer infront but don't let you click em, sometimes highlighting two at the same time
Funny, wild jumping when collapsing/reopening categories
(Maybe bug, maybe not?) Changing the Debug button to e.g. `G` and pressing it, opens the debug menu _and_ also shows chunkborders. Just mentioning here just in case (maybe some later check, e.g. highlighting the button or sth would make sense, anyway, probably not important)

(Haven't messed around much with things like Filters n such yet tho)

Anyways, pretty nice progress so far :D

**Comment by BioTechproject27** [comment]
Hey so, I just saw this:
https://www.youtube.com/watch?v=fTd6kTMLkuQ
(the most important part is probably at 4:19)
According to this, it seems like **client-side crawling is now also valid server-side**,
so I thought maybe one could implement crawling as a keybind, it should work by just the client telling itself that it's crawling(?), and that should enable one to actually be able to crawl (at least in single player, though that may be glitchy due to server zap-back, so a 'warning' at first would make sense that it's just experimental; And it does give an unfair advantage if used on servers [though I did feel that should be a vanilla feature], so it should come with a warning that this could get you banned [or not work at all + be glitchy])

This is an additional feature idea, so if anything a low priority for now, I'd like to see your thoughts on it sometime, I'd say it's definitely super interesting :D

**Comment by agent-LuluDodo** [comment]
I have finally fixed most of the bugs, it's basically just mod support and an api I need to write, before the mod is getting released 🥳

And I'm not going to be adding client-side crawling for two reasons:
1. It's not possible in vanilla per key bind
2. As the video showcases, a tiny (1/10000 of a block inconsistency in hit box calculations is abused and thus makes crawling only server side if you're basically under a block, this might still be useful but seems more like a bug than a feature)

**Comment by BioTechproject27** [comment]
Nice nice :D

> And I'm not going to be adding client-side crawling

Fair enough fair enough, if anything it should be properly supported server side as well
Just thought it was kinda cool xdd
